Add CRIU version check for checkpoint and restore

adrianreber opened the pull request:
As the checkpoint/restore implementation in Podman requires at least CRIU 3.11, this vendors in the Go bindings of CRIU and uses these bindings to check the version of the installed CRIU binary.

TomSweeneyRedHat commented:
/ok-to-test

mheon commented on libpod/container_internal_linux.go:
```diff
@@ -355,6 +360,10 @@ func (c *Container) checkpoint(ctx context.Context, keep bool) (err error) {
 func (c *Container) restore(ctx context.Context, keep bool) (err error) {
 
+	if !criu.CheckForCriu() {
+		return errors.Errorf("checkpointing a container requires at least CRIU 3.11")
```
Change to "restoring a container" instead of "checkpointing a container"

adrianreber replied:
Ah, copy/paste. Will fix.

mheon commented on pkg/criu/criu.go:
```diff
@@ -0,0 +1,17 @@
+func CheckForCriu() bool {
+	c := criu.MakeCriu()
+	// For Podman at least CRIU 3.11 is required
+	result, err := c.IsCriuAtLeast(31100)
```
Nit - version number would probably be better as a package constant

mheon commented:
/approve

mheon commented:
@adrianreber We just merged something that should fix CI, mind rebasing?

adrianreber commented:
@mheon thanks for the review. I changed the things you pointed out and force pushed.

I now get the following if CRIU is too old:

```
#  podman container checkpoint --keep podman-criu-test 
failed to checkpoint container 91b1f32f7: checkpointing a container requires at least CRIU 31100
```

baude commented:
/retest images

baude commented:
/test images

baude commented:
bot, retest this please

mheon commented:
/retest ci/prow/images

baude commented:
/test images

mheon commented:
/test images

adrianreber commented:
Yes, the vendoring is not correct, need to fix that.

adrianreber commented:
I just saw that CRIU is no longer installed in the new test environment. All checkpoint/restore test cases are now skipped.

rhatdan commented:
/approve

rhatdan commented:
lgtm

mheon commented:
Hm. I'll talk to Chris and Brent and see about getting CRIU on our test VMs
in Cirrus

On Fri, Oct 12, 2018, 05:41 Daniel J Walsh <notifications@github.com> wrote:

> lgtm
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/containers/libpod/pull/1627#issuecomment-429268278>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHYHCAD1HmI1I7gvExCJohbA3yJ3avJJks5ukGPigaJpZM4XXC3o>
> .
>

adrianreber commented:
> Hm. I'll talk to Chris and Brent and see about getting CRIU on our test VMs in Cirrus

The necessary steps are still in the Dockerfile.

mheon commented:
A lot of our tests are now on bare VMs, so those won't run the Dockerfile -
we'll need it added there. Shouldn't be that bad.

On Fri, Oct 12, 2018, 09:09 Adrian Reber <notifications@github.com> wrote:

> Hm. I'll talk to Chris and Brent and see about getting CRIU on our test
> VMs in Cirrus
>
> The necessary steps are still in the Dockerfile.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/containers/libpod/pull/1627#issuecomment-429319661>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHYHCARIVmp0imwDwu96sbFXlyVWHXUFks5ukJRxgaJpZM4XXC3o>
> .
>

baude commented:
For F28 in PAPR, if the packaged version is sufficient, then an edit here would do it: -> https://github.com/containers/libpod/blob/master/.papr.yml#L112

@cevich can you say where we can add additional packages for cirrus?

cevich commented:
@baude currently all package installing happens at image-build time in ``contrib/cirrus/packer/*.sh`` for the various distros.  However, if this is something that you expect to update frequently (more than once a week or so), it may make more sense to install it at runtime via the distro-specific ``case`` blocks in ``contrib/cirrus/setup_environment.sh``.  If neither makes sense, let me know.

adrianreber commented:
CRIU would not need to be updated often. Only if a new feature is required, which Podman needs.

Currently Podman needs CRIU from git (that's what I added to the docker file when travis testing was still done). Once CRIU 3.11 is released (probably in the next few weeks), we can just install the Fedora package (for the Fedora tests), once I have updated the package there.

I would like to see checkpoint/restore testing enabled again in Podman before CRIU 3.11 is released to make sure that checkpoint/restore does not brake, that would mean installation from git.

But if you say that this is too much effort we can also wait until CRIU 3.11 is released and is part of Fedora. Whatever works best. Let me know if there is anything I can do to help here.

mheon commented:
I'm fine with building from Git - we have a few other dependencies we grab this way (runc, for example, we're building a specified commit there).

This will need changes to the Cirrus and Papr scripts and the Dockerfile to install from source, but I don't think they should be that bad

cevich commented:
> (that's what I added to the docker file when travis testing was still done)

Oh, that makes it even easier.  I can just copy-paste what you did for the Dockerfile, [something like this for cirrus.](https://paste.fedoraproject.org/paste/X7aZmU6zAuM8I1fovotVRQ) super easy.

cevich commented:
Okay done: #1660

I added it for all distros, but should it be just for Ubuntu?

adrianreber commented:
@cevich thanks, looks perfect. CRIU and podman checkpoint/restore should work on all distros.

cevich commented:
@adrianreber excellent, okay I'l carry on with that then, thanks.

adrianreber commented:
/retest

adrianreber commented:
All automated tests are now green (except cirrus ci which still is Pending)
